File: macos/Classes/Models/WhisperSummary.swift
import Foundation

struct WhisperSummary {
    var tSampleUs: Int64
    var nSample: Int32
    var tEncodeUs: Int64
    var nEncode: Int32
    var tDecodeUs: Int64
    var nDecode: Int32
    var tPromptUs: Int64
    var nPrompt: Int32
    var tStartUs: Int64
    var tEndUs: Int64
    var nFailP: Int32
    var nFailH: Int32
    var tLoadUs: Int64
    var tMelUs: Int64
    
    init(summary: whisper_summary) {
        self.tSampleUs = summary.t_sample_us
        self.nSample = summary.n_sample
        self.tEncodeUs = summary.t_encode_us
        self.nEncode = summary.n_encode
        self.tDecodeUs = summary.t_decode_us
        self.nDecode = summary.n_decode
        self.tPromptUs = summary.t_prompt_us
        self.nPrompt = summary.n_prompt
        self.tStartUs = summary.t_start_us
        self.tEndUs = summary.t_end_us
        self.nFailP = summary.n_fail_p
        self.nFailH = summary.n_fail_h
        self.tLoadUs = summary.t_load_us
        self.tMelUs = summary.t_mel_us
    }
    
    func toDictionary() -> [String: Any] {
        return [
            "tSampleUs": self.tSampleUs,
            "nSample": self.nSample,
            "tEncodeUs": self.tEncodeUs,
            "nEncode": self.nEncode,
            "tDecodeUs": self.tDecodeUs,
            "nDecode": self.nDecode,
            "tPromptUs": self.tPromptUs,
            "nPrompt": self.nPrompt,
            "tStartUs": self.tStartUs,
            "tEndUs": self.tEndUs,
            "nFailP": self.nFailP,
            "nFailH": self.nFailH,
            "tLoadUs": self.tLoadUs,
            "tMelUs": self.tMelUs
        ]
    }
}

File: lib/src/models/whisper_summary/whisper_summary.dart
import 'package:whisper_cpp/whisper_cpp.dart';

class WhisperSummary {
  int tSampleUs;
  int nSample;
  int tEncodeUs;
  int nEncode;
  int tDecodeUs;
  int nDecode;
  int tPromptUs;
  int nPrompt;
  int tStartUs;
  int tEndUs;
  int nFailP;
  int nFailH;
  int tLoadUs;
  int tMelUs;

  WhisperSummary({
    required this.tSampleUs,
    required this.nSample,
    required this.tEncodeUs,
    required this.nEncode,
    required this.tDecodeUs,
    required this.nDecode,
    required this.tPromptUs,
    required this.nPrompt,
    required this.tStartUs,
    required this.tEndUs,
    required this.nFailP,
    required this.nFailH,
    required this.tLoadUs,
    required this.tMelUs,
  });

  double get sampleTimePerRun => tSampleUs / 1000.0 / nSample;
  double get encodeTimePerRun => tEncodeUs / 1000.0 / nEncode;
  double get decodeTimePerRun => tDecodeUs / 1000.0 / nDecode;
  double get promptTimePerRun => tPromptUs / 1000.0 / nPrompt;
  double get totalTime => (tEndUs - tStartUs) / 1000.0;

  double get loadTime => tLoadUs / 1000.0;
  String get fallbacks => '$nFailP p / $nFailH h';
  double get melTime => tMelUs / 1000.0;

  factory WhisperSummary.fromJson(Map<Object?, Object?>? data) {
    Map<String, dynamic> json = Map<String, dynamic>.from(data as Map);

    return WhisperSummary(
      tSampleUs: IntHandler.parse(json['tSampleUs']),
      nSample: IntHandler.parse(json['nSample']),
      tEncodeUs: IntHandler.parse(json['tEncodeUs']),
      nEncode: IntHandler.parse(json['nEncode']),
      tDecodeUs: IntHandler.parse(json['tDecodeUs']),
      nDecode: IntHandler.parse(json['nDecode']),
      tPromptUs: IntHandler.parse(json['tPromptUs']),
      nPrompt: IntHandler.parse(json['nPrompt']),
      tStartUs: IntHandler.parse(json['tStartUs']),
      tEndUs: IntHandler.parse(json['tEndUs']),
      nFailP: IntHandler.parse(json['nFailP']),
      nFailH: IntHandler.parse(json['nFailH']),
      tLoadUs: IntHandler.parse(json['tLoadUs']),
      tMelUs: IntHandler.parse(json['tMelUs']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'tSampleUs': tSampleUs,
      'nSample': nSample,
      'tEncodeUs': tEncodeUs,
      'nEncode': nEncode,
      'tDecodeUs': tDecodeUs,
      'nDecode': nDecode,
      'tPromptUs': tPromptUs,
      'nPrompt': nPrompt,
      'tStartUs': tStartUs,
      'tEndUs': tEndUs,
      'nFailP': nFailP,
      'nFailH': nFailH,
      'tLoadUs': tLoadUs,
      'tMelUs': tMelUs,
    };
  }
}
